capstone_project/src/main/java/com/google/sps/servlets/MapServlet.java

```diff
@@ -28,52 +38,63 @@
-                "https://news.uci.edu/2020/06/18/uci-podcast-jessica-millward-on-the-meaning-and-importance-of-juneteenth/"));
-    School uwb = new School("UWB", 47.758821, -122.190725, uwbNews);
-    schools.add(uwb);
-  }
+    // Load Schools from Datastore.
+    Query query = new Query("School");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
 
-  /*Add some hard-coded School objects to the array of schools that will later be converted into markers on the map.*/
-  public MapServlet() {
-    addUCIToSchools();
-    addUWToSchools();
-    addUWBToSchools();
-  }
+    for (Entity entity : results.asIterable()) {
+      String schoolName = (String) entity.getProperty("name");
+      double schoolLatitude = (double) entity.getProperty("latitude");
+      double schoolLongitude = (double) entity.getProperty("longitude");
+      School retrievedSchool = new School(schoolName, schoolLatitude, schoolLongitude);
+      schoolsSet.add(retrievedSchool);
+    }
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Send the School objects retrieved from Datastore.
+    ArrayList<School> schools = new ArrayList<School>(schoolsSet);
     response.setContentType("application/json;");
     Gson gson = new Gson();
     response.getWriter().println(gson.toJson(schools));
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Initialize datastore object.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get the input from the school entry form.
+    String userName = getParameter(request, "name-input", "");
+    double userLatitude = Double.parseDouble(getParameter(request, "latitude-input", ""));
+    double userLongitude = Double.parseDouble(getParameter(request, "longitude-input", ""));
+
+    // Check if the School is already in Datastore.
+    Filter nameFilter = new FilterPredicate("name", FilterOperator.EQUAL, userName);
+    Query query = new Query("School").setFilter(nameFilter);
+    PreparedQuery results = datastore.prepare(query);
+
+    // If it is not, create a schoolEntity for the School and add it to Datastore.
+    if (results.countEntities(FetchOptions.Builder.withDefaults()) == 0) {
+      Entity schoolEntity = new Entity("School");
+      schoolEntity.setProperty("name", userName);
+      schoolEntity.setProperty("latitude", userLatitude);
+      schoolEntity.setProperty("longitude", userLongitude);
+      datastore.put(schoolEntity);
+    }
+
+    // Respond with the result.
+    response.sendRedirect("pages/maps.html");
+  }
+
+  /** @return the request parameter, or the default value if not specified */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
```

Comment:
What do you plan on doing with this method? This might be carry over from the starter project, I feel like I remember something similar but not entirely sure. As of right now, you pass in an empty string to defaultValue, so just curious if you have plans for this method!

Reply:
@lyndapham Thanks for catching this, I actually decided to delete the method, since there's no need to use the default value.